File: src/CommandRegistry.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Text;

namespace InoCLI
{
   // ============================================================
   /// <summary>
   /// Scans, stores, and resolves CLI commands.
   /// </summary>
   // ============================================================
   public class CommandRegistry
   {

   #region Fields

      private readonly Dictionary<string, CommandInfo> commands = new Dictionary<string, CommandInfo>();

      private int maxDepth = 0;

      private List<string> cachedRoots = null;
      private string       cachedHelp  = null;

   #endregion

   #region Initialize

      // ------------------------------------------------------------
      /// <summary>
      /// Scans assemblies and registers all [CLICommand] methods.
      /// </summary>
      // ------------------------------------------------------------
      public void Initialize(params Assembly[] assemblies)
      {
         Initialize((IEnumerable<Assembly>)assemblies);
      }

      // ------------------------------------------------------------
      /// <summary>
      /// Scans assemblies and registers all [CLICommand] methods.
      /// </summary>
      // ------------------------------------------------------------
      public void Initialize(IEnumerable<Assembly> assemblies)
      {
         commands.Clear();

         maxDepth    = 0;
         cachedRoots = null;
         cachedHelp  = null;

         var discovered = CommandScanner.ScanAll(assemblies);

         foreach (var info in discovered)
         {
            commands[info.Key] = info;

            if (info.Path.Length > maxDepth)
            {
               maxDepth = info.Path.Length;
            }
         }
      }

   #endregion

   #region Resolve

      // ----------------------------------------------------------------------
      /// <summary>
      /// <br/> Matches positionals against registered command paths.
      /// <br/> Returns the matched command and a new CommandArgs
      /// <br/> with the path segments removed from positionals.
      /// </summary>
      // ----------------------------------------------------------------------
      public (CommandInfo info, CommandArgs args) Resolve(CommandArgs parsed)
      {
         // Greedy match: try longest path first
         int depth = Math.Min(parsed.Positionals.Count, maxDepth);

         for (; depth >= 1; depth--)
         {
            string key = string.Join(".", parsed.Positionals.GetRange(0, depth));

            if (commands.TryGetValue(key, out var info))
            {
               var args = new CommandArgs
               {
                  Positionals = parsed.Positionals.GetRange(depth, parsed.Positionals.Count - depth),
                  Optionals   = parsed.Optionals
               };

               return (info, args);
            }
         }

         string attempted = parsed.Positionals.Count > 0 ? parsed.Positionals[0] : "(empty)";

         throw new ArgumentException($"Unknown command: {attempted}");
      }

   #endregion

   #region Query

      // ------------------------------------------------------------
      /// <summary>
      /// Finds a command by exact path, or null if not found.
      /// </summary>
      // ------------------------------------------------------------
      public CommandInfo Find(params string[] path)
      {
         string key = string.Join(".", path);

         commands.TryGetValue(key, out var info);

         return info;
      }

      // ------------------------------------------------------------
      /// <summary>
      /// Returns all registered commands.
      /// </summary>
      // ------------------------------------------------------------
      public List<CommandInfo> GetAll()
      {
         return commands.Values.ToList();
      }

      // ------------------------------------------------------------
      /// <summary>
      /// Returns distinct root path segments (Path[0]). Cached.
      /// </summary>
      // ------------------------------------------------------------
      public List<string> GetRoots()
      {
         if (cachedRoots != null)
         {
            return cachedRoots;
         }

         var roots = new HashSet<string>();

         foreach (var info in commands.Values)
         {
            if (info.Path.Length > 0)
            {
               roots.Add(info.Path[0]);
            }
         }

         cachedRoots = roots.ToList();

         return cachedRoots;
      }

   #endregion

   #region Help

      // ------------------------------------------------------------
      /// <summary>
      /// Generates help text for all registered commands. Cached.
      /// </summary>
      // ------------------------------------------------------------
      public string GetHelp()
      {
         if (cachedHelp != null)
         {
            return cachedHelp;
         }

         var sb = new StringBuilder();

         var sorted = commands.Values.OrderBy(c => c.Key).ToList();

         foreach (var info in sorted)
         {
            sb.Append("  ");
            sb.Append(info.Key.PadRight(30));

            if (!string.IsNullOrEmpty(info.Description))
            {
               sb.Append(info.Description);
            }

            sb.AppendLine();
         }

         cachedHelp = sb.ToString();

         return cachedHelp;
      }

      // ----------------------------------------------------------------------
      /// <summary>
      /// <br/> Generates help text for commands under a specific path prefix.
      /// <br/> e.g. GetHelp("scene") lists all scene.* commands.
      /// </summary>
      // ----------------------------------------------------------------------
      public string GetHelp(params string[] path)
      {
         string prefix = string.Join(".", path);

         var sb = new StringBuilder();

         var matched = commands.Values
            .Where(c => c.Key.StartsWith(prefix, StringComparison.Ordinal))
            .OrderBy(c => c.Key)
            .ToList();

         foreach (var info in matched)
         {
            // Show path relative to the prefix
            string relative = info.Key.Length > prefix.Length
               ? info.Key.Substring(prefix.Length + 1)
               : info.Key;

            sb.Append("  ");
            sb.Append(relative.PadRight(30));

            if (!string.IsNullOrEmpty(info.Description))
            {
               sb.Append(info.Description);
            }

            sb.AppendLine();
         }

         return sb.ToString();
      }

   #endregion

   }
}

File: rust/inocli/src/command_registry.rs
use std::collections::{HashMap, HashSet};

use crate::{CliError, CommandArgs, CommandInfo};

pub struct CommandRegistry {
    commands: HashMap<String, CommandInfo>,
    max_depth: usize,
}

impl CommandRegistry {
    pub fn new() -> Self {
        CommandRegistry {
            commands: HashMap::new(),
            max_depth: 0,
        }
    }

    pub fn register(&mut self, info: CommandInfo) -> &mut Self {
        if info.path.len() > self.max_depth {
            self.max_depth = info.path.len();
        }
        self.commands.insert(info.key.clone(), info);
        self
    }

    /// Greedy longest-path match. Strips matched path segments from positionals.
    pub fn resolve(&self, args: &CommandArgs) -> Result<(&CommandInfo, CommandArgs), CliError> {
        let depth = args.positionals.len().min(self.max_depth);

        for d in (1..=depth).rev() {
            let key = args.positionals[..d].join(".");
            if let Some(info) = self.commands.get(&key) {
                let remaining = CommandArgs {
                    positionals: args.positionals[d..].to_vec(),
                    optionals: args.optionals.clone(),
                };
                return Ok((info, remaining));
            }
        }

        let attempted = args.positionals.first().map_or("(empty)", |s| s.as_str());
        Err(CliError::UnknownCommand(attempted.to_string()))
    }

    pub fn find(&self, path: &[&str]) -> Option<&CommandInfo> {
        self.commands.get(&path.join("."))
    }

    pub fn get_all(&self) -> Vec<&CommandInfo> {
        self.commands.values().collect()
    }

    pub fn get_roots(&self) -> Vec<&str> {
        let mut roots: HashSet<&str> = HashSet::new();
        for info in self.commands.values() {
            if let Some(first) = info.path.first() {
                roots.insert(first.as_str());
            }
        }
        roots.into_iter().collect()
    }

    pub fn get_help(&self) -> String {
        let mut sorted: Vec<&CommandInfo> = self.commands.values().collect();
        sorted.sort_by(|a, b| a.key.cmp(&b.key));

        let mut help = String::new();
        for info in sorted {
            help.push_str(&format!("  {:<30}", info.key));
            if !info.description.is_empty() {
                help.push_str(&info.description);
            }
            help.push('\n');
        }
        help
    }

    pub fn get_help_for(&self, prefix: &[&str]) -> String {
        let prefix_key = prefix.join(".");
        let mut matched: Vec<&CommandInfo> = self
            .commands
            .values()
            .filter(|c| c.key.starts_with(&prefix_key))
            .collect();
        matched.sort_by(|a, b| a.key.cmp(&b.key));

        let mut help = String::new();
        for info in matched {
            let relative = if info.key.len() > prefix_key.len() {
                &info.key[prefix_key.len() + 1..]
            } else {
                &info.key
            };
            help.push_str(&format!("  {:<30}", relative));
            if !info.description.is_empty() {
                help.push_str(&info.description);
            }
            help.push('\n');
        }
        help
    }
}

impl Default for CommandRegistry {
    fn default() -> Self {
        Self::new()
    }
}
